### ISSUE TITLE
Re-inserting identical entity throws 'InvalidArgument' and not 'AlreadyExists'

### ISSUE THREAD

**Issue by Dafeesh** [GoogleCloudPlatform/google-cloud-dotnet#2453]
When attempting to insert an identical entity twice, the expected GrpcException status code has been: 'Grpc.Core.StatusCode.AlreadyExists'. But sometime along the way this exception has changed to throw 'Grpc.Core.StatusCode.InvalidArgument'. 

When attempting to add an entity that already exists in the DB it would make the most sense to throw 'AlreadyExists' and use 'InvalidArgument' for strictly key formatting issues, etc.

How to reproduce:
Run "gcloud beta emulators datastore start"
Attempt to insert two identical entities: { _db.Insert(myEntity); _db.Insert(myEntity); }
Catch exception on the second call and observe the following exception: 'RpcException: Status(StatusCode=InvalidArgument, Detail="entity already exists")'

**Comment by jskeet** [comment]
I'll have a look at this on Tuesday. (Monday is a public holiday in the UK.) If it only happens in the emulator, it's probably an emulator bug. (It's very unlikely that this is a client library issue. We don't do much with errors other than retry some operations in some cases.)

**Comment by jskeet** [comment]
I can't reproduce this, either against the emulator or the production environment. Here's an example which attempts to reproduce it:

```csharp
using Google.Cloud.Datastore.V1;
using Grpc.Core;
using System;

class Program
{
    static void Main(string[] args)
    {
        string emulatorHost = Environment.GetEnvironmentVariable("DATASTORE_EMULATOR_HOST");
        string projectId = Environment.GetEnvironmentVariable("DATASTORE_PROJECT_ID");
        DatastoreClient client = string.IsNullOrEmpty(emulatorHost)
            ? DatastoreClient.Create()
            : DatastoreClient.Create(new Channel(emulatorHost, ChannelCredentials.Insecure));
        var db = DatastoreDb.Create(projectId, "Issue2453", client);

        var entity = new Entity
        {
            Key = db.CreateKeyFactory("demo").CreateKey(Guid.NewGuid().ToString()),
            Properties =
            {
                ["text"] = "Demo"
            }
        };

        Console.WriteLine("First insertion");
        db.Insert(entity);
        Console.WriteLine("Second insertion");
        db.Insert(entity);
    }
}
```

In both cases, the first insertion succeeds; the second fails with:

> Unhandled Exception: Grpc.Core.RpcException: Status(StatusCode=AlreadyExists, Detail="entity already exists")

Set the `DATASTORE_PROJECT_ID` environment variable, and then either set or clear `DATASTORE_EMULATOR_HOST` based on whether you want to test against the emulator or not.

Could you try that code in your environment?

**Comment by Dafeesh** [comment]
I'm confused how I got to this point but I was somehow running an extremely old version of the datastore emulator even after updating. After cleaning out my installation and installing the SDK again I'm now up to date and it's giving me the expected result similar to what you posted above. Thank you for investigating. 

I found that this specific issue was fixed in datastore emulator version 1.4.
""
Released Cloud Datastore Emulator version 1.4.0.
   The emulator now returns error codes that match the Cloud Datastore service for two scenarios: inserts of entities that already exist and updates to ones that don't.
""
